server/entity/splash_potion.go

```diff
@@ -0,0 +1,248 @@
+func (s *SplashPotion) Name() string {
+	return "Splash Potion"
+}
+
+// EncodeEntity ...
+func (s *SplashPotion) EncodeEntity() string {
+	return "minecraft:splash_potion"
+}
+
+// AABB ...
+func (s *SplashPotion) AABB() physics.AABB {
+	return physics.NewAABB(mgl64.Vec3{-0.125, 0, -0.125}, mgl64.Vec3{0.125, 0.25, 0.125})
+}
+
+// SetType ...
+func (s *SplashPotion) SetType(t potion.Potion) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.t = t
+	for _, v := range s.e.World().Viewers(s.pos) {
+		v.ViewEntityState(s.e)
+	}
+}
+
+// Type ...
+func (s *SplashPotion) Type() potion.Potion {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.t
+}
+
+// Rotation ...
+func (s *SplashPotion) Rotation() (float64, float64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.yaw, s.pitch
+}
+
+// splashable represents an entity that can be splashed by a potion.
+type splashable interface {
+	Living
+	Eyed
+
+	// AddEffect adds a specific effect to the entity that implements this interface.
+	AddEffect(e effect.Effect)
+}
+
+// Tick ...
+func (s *SplashPotion) Tick(current int64) {
+	if s.closeNextTick {
+		_ = s.Close()
+		return
+	}
+	s.mu.Lock()
+	m, result := s.c.TickMovement(s, s.pos, s.vel, s.yaw, s.pitch, s.ignores)
+	s.pos, s.vel, s.yaw, s.pitch = m.pos, m.vel, m.yaw, m.pitch
+	s.mu.Unlock()
+
+	s.ticksLived++
+	m.Send()
+
+	if m.pos[1] < cube.MinY && current%10 == 0 {
+		s.closeNextTick = true
+		return
+	}
+
+	if result != nil {
+		w := s.World()
+		aabb := s.AABB().Translate(m.pos)
+
+		effects := s.t.Effects()
+		hasEffects := len(effects) > 0
+
+		colour := color.RGBA{R: 0x38, G: 0x5d, B: 0xc6, A: 0xff}
+
+		if hasEffects {
+			colour, _ = effect.ResultingColour(effects)
+
+			ignore := func(entity world.Entity) bool {
+				_, canSplash := entity.(splashable)
+				return !canSplash || entity == s
+			}
+
+			for _, otherEntity := range w.EntitiesWithin(aabb.GrowVec3(mgl64.Vec3{8.25, 4.25, 8.25}), ignore) {
+				splashEntity := otherEntity.(splashable)
+				if !splashEntity.AABB().Translate(splashEntity.Position()).IntersectsWith(aabb.GrowVec3(mgl64.Vec3{4.125, 2.125, 4.125})) {
+					continue
+				}
+
+				distance := world.Distance(EyePosition(splashEntity), m.pos)
+				if distance > 4 {
+					continue
+				}
+
+				distanceMultiplier := 1 - distance/4
+				if entityResult, ok := result.(trace.EntityResult); ok && entityResult.Entity() == otherEntity {
+					distanceMultiplier = 1
+				}
+
+				for _, eff := range effects {
+					if _, ok := eff.Type().(effect.LastingType); !ok {
+						splashEntity.AddEffect(eff.WithPotency(distanceMultiplier))
+						continue
+					}
+
+					distanceAccountedDuration := time.Duration(float64(eff.Duration().Milliseconds())*0.75*distanceMultiplier) * time.Millisecond
+					if distanceAccountedDuration < time.Second {
+						continue
+					}
+					splashEntity.AddEffect(eff.WithDuration(distanceAccountedDuration))
+				}
+			}
+		} else if blockResult, ok := result.(trace.BlockResult); ok && s.t.Equals(potion.Water()) {
```

Comment:
These water splash potions also deal damage to endermen, blazes, striders and snow golems. Could you add a TODO for that?